### ISSUE TITLE
Improve posthog-plugins-ingestion's throughput

### ISSUE THREAD

**Issue by xvello** [PostHog/posthog#15518]
## In what situation are you experiencing subpar performance?

- The `posthog-plugins-ingestion` deploy on prod-us is getting too big for comfort
- The less partitions are assigned to a pod, the lower its throughput gets, which breaks our `input = pods * throughput/pod` capacity planning equation: we think we have capacity to grow, but upscaling does not help

## Current situation

This workload still uses the `kafkajs` consumer. For performance and operability reasons, our goal this quarter is to move to the `rdkafka` consumer (producer side already done for all workloads).

- The kafkajs consumer handles batch fetching and calls `eachBatchIngestion` with each batch.
- `eachBatchIngestion` splits that batch into contiguous micro-batches with `groupIntoBatchesIngestion`. Events within a micro-batch are processed in parallel, so enforcing that events from a given `distinct_id` are processed sequentially creates really small micro-batches *(an average of 2.5 events only!)* that don't saturate one CPU core
- For each event in the micro-batch, `runEventPipeline` is called concurrently. Most of that processing is CPU-bound (as processing apps on Cloud are not allowed to make network calls), with the exception of:
  - `processPersonsStep` does several async IO to PG and Redis to read and update Persons records
  - `createEventStep` (the last processing step for analytics events) writes the processed event to the Kafka queue, but also await for the write ACK. This is accounting for a huge amount of async IO
- Once all events in a micro-batch are processed, consumer offsets are committed (async IO) to reduce the risk of duplicates.

That many async IO make it hard to efficiently use the allocated CPU. On Cloud, we mitigate this with the `KAFKA_PARTITIONS_CONSUMED_CONCURRENTLY` parameter, that makes `kafkajs` run several `eachBatchIngestion` loops (currently 5) with batches from different partitions. But that mitigation is why our efficiency goes down when we upscale the workload: the pods are waiting on async IO, that basically become blocking IO.

While moving the mostly-CPU-bound `posthog-recordings-ingestion` workload to `rdkafka` gave us a >10x throughput increase, moving the `posthog-plugins-ingestion` workload to the `rdkafka` consumer yielded a 2x throughput loss because our new code runs a single `eachBatchIngestion` loop. For the `rdkafka` transition to be successful, we need to reduce the amount of async IO blocking the main nodejs loop and saturate our CPU better.

## Plan forward

Several iterative changes to be implemented:

- One of `rdkafka`'s performance trick is aggressive write batching. While `runEventPipeline` should write to Kafka as soon as possible, it should not wait for the write ACK before processing another event. Because `rdkafka` is handling retries behind the scenes, we can assume a queued event will be produced. We just want to make sure we don't process faster than we can write, by waiting for the Kafka ACK once a batch. We will change `runEventPipeline`'s signature to move the ACK wait & offset commit up to the toplevel `eachBatchIngestion` loop. This will significantly reduce the amount of async IO we're doing. It will increase the risk of duplicates if the pod dies, but still within the bounds of a small 500-message batch
- Once we commit once per batch, we don't need micro-batches sliced by `groupIntoBatchesIngestion` to be contiguous offsets anymore. We will rework the micro-batching logic to achieve higher concurrency, even with a single partition:
  - Events from the same `distinct_id` will be grouped into micro-batches
  - Events in one micro-batch will be processed sequentially
  - A configurable number (5 to 10?) of parallel loops will consume these micro-batches, from bigger to smaller via a weighted queue
  - Events that fail due to `DependencyUnavailableError` are retried until they pass, blocking the consumer's progress
  - Once all micro-batches are properly processed, we wait on ACK, commit offsets on all assigned partitions, and start processing the next batch
- `rdkafka` handles pre-fetching of events, making the consumer read almost non-blocking. Waiting on ACK and commiting offsets are delegated to helper C threads, but are still blocking the nodejs main loop. We can do these operations on the `N-1` batch while we process batch `N`, hopefully squeezing out a bit more performance out of the loop.

## Going further: improving `processPersonsStep`s performance

`processPersonsStep` has known data quality and performance issues, and its data flow is hard to follow. After studying it with @tiina303, I recommend the following improvements: 

* for the creation: use `insert .... on conflict $index do nothing`, instead of swallowing the error. Then any error from that query should trigger a retry of the event
* partial updates of the person's `properties` jsonb, instead of sending the whole `properties` json again: better perf (ser/deser) + don’t swallow new properties when several distinct_id are updating the same person
* optimistic writes with `update where version = N`. if no row updated, it's a race on either update or delete, bust cache and retry
* consider making use of the fact that we're micro-batching event from the same distinct_id together: optimistic writes would allow us to reuse the `Person` object between events of a single micro-batch

## Going further: consider reintroducing multi-threading to use more CPU cores per consumer

With the rdkafka operations delegated to helper threads, we'll hopefully manage to saturate the main nodejs thread with the `eachBatchIngestion` loop. Once that's the case, profiling will help us squeeze out more performance, up to a point.

The next leap will require us to use multiple cores per consumer, which we could achieve with [node worker threads](https://nodejs.org/api/worker_threads.html), either:

  - slicing horizontally, by running micro-batch consuming loops in different threads
  - slicing vertically, by running different ingestion steps in different threads, each processing each of the micro-batches sequentially.

This is out of scope for now, but something to keep in mind for Q3/Q4.

**Comment by tiina303** [comment]
Questions / notes:
1. `processPersonsStep` - don't we wait for kafka writes there? 
2. Note: processing micro-batches of different distinct_ids in parallel could cause problems for some users, e.g. if a user sends us a single batch of events that have events in this order, then currently it's guaranteed to work, but after this change
    1. alias(a, b)
    2. alias(b, c)
    3. alias(c, d)  -- if this is ran before (2), then (2) will fail as c is already is_identified. This said it's not how one should use PostHog (I'm fairly confident 🤔 might be good to check in #ask-posthog-anything too), you should always only identify towards a single distinct_id, so we can ask them to change what they are doing. But just for us to be aware of this.
    Separately I'll note that person properties could updated out of order across distinctIds (we've already made the assumption that person props shouldn't be updated often and we can just reinforce that).
1. `personProcessStep` - for creation failures the optimal thing would be to return the already existing row (or fetch it) and update properties if needed (we don't need to retry if it's due to the row already existing). On failure to write for other reasons yes we should retry.
1. `personProcessStep` updates: we only need one of optimistic updates or partial updates or am I missing something?
1. `processPersonStep` "consider making use of the fact that we're micro-batching event from the same distinct_id together: optimistic writes would allow us to reuse the Person object between events of a single micro-batch" ... interesting idea, Redis cache across everything could be an alternative to consider, but 🤷‍♀️  In any case this would be one of the later improvements to make compared to others.
1. with further multi-threading we need to be careful about sequential processing for the same distinct ID or better yet figure out how to get rid of that constraint, so we can process everything in parallel and simplify all the things.

**Comment by mano233** [comment]
any update?we now have over 150k events per minute, and adding more replicas of the plugins-server has not increased the consumption rate of the system.

**Comment by mano233** [comment]
> Pipeline team's Q1 objective: "Person model doesn't slow down the pipeline" here: https://posthog.com/handbook/small-teams/pipeline#objectives-q1-2024 We'll be discussing the options/plans in the offsite end of January

good job👌

**Comment by xvello** [comment]
Closing this issue as all but the person model changes have been successfully implemented, leading to one order of magnitude improvements in throughput. Person model is now our main bottleneck and a Q1 priority.

@mano233 The design goal of our hobby deploy is a set-and-forget setup with no maintenance, and that’s why it’s designed [with low volume targets](https://posthog.com/docs/self-host/open-source/disclaimer) because, in our experience, higher volumes will require tuning and maintenance of the data stores that most users would not be able to safely achieve.

I am happy to learn that our ongoing throughput optimization work translates so well to the hobby deploy, but you seem to be operating way over the volumes that our hobby topology is designed for. We will not support horizontal scaling of the hobby deploy, nor help users who attempt it.
